pkg/cmd/pulumi/templates/cloud.go

```diff
@@ -29,18 +29,127 @@ import (
 	pkgWorkspace "github.com/pulumi/pulumi/pkg/v3/workspace"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/env"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/registry"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
 )
 
+func (s *Source) getCloudTemplates(
+	ctx context.Context, templateName string,
+	wg *sync.WaitGroup, e env.Env,
+) {
+	if !e.GetBool(env.DisableRegistryResolve) && e.GetBool(env.Experimental) {
+		s.getRegistryTemplates(ctx, e, templateName)
+		return
+	}
+
+	// Use the old org templates based API.
+	//
+	// This path can be removed when we are confident in registry resolution. We will
+	// always need to maintain a way to access templates without the service, but we
+	// should only need to maintain one way to access templates through the service.
+	s.getOrgTemplates(ctx, templateName, wg, e)
+}
+
+func (s *Source) getRegistryTemplates(ctx context.Context, e env.Env, templateName string) {
+	r := registry.NewOnDemandRegistry(func() (registry.Registry, error) {
+		b, err := cmdBackend.NonInteractiveCurrentBackend(
+			ctx, pkgWorkspace.Instance, cmdBackend.DefaultLoginManager, nil,
+		)
+		if err == nil && b != nil {
+			return b.GetReadOnlyCloudRegistry(), nil
+		}
+		if b == nil || errors.Is(err, backenderr.ErrLoginRequired) {
+			return unauthenticatedregistry.New(cmdutil.Diag(), e), nil
+		}
+		return nil, fmt.Errorf("could not get registry backend: %w", err)
+	})
+
+	// Since the templates names displayed here differ from the template names
+	// returned from ListTemplates for VCS backed templates, we need to fetch
+	// all templates and then filter manually.
+	var nameFilter *string
+
+	for template, err := range r.ListTemplates(ctx, nameFilter) {
+		if err != nil {
+			s.addError(fmt.Errorf("could not get template: %w", err))
+			return
+		}
+
+		if template.Source == "github" && strings.HasPrefix(template.Name, "pulumi/templates/") {
+			// These template are maintained using https://github.com/pulumi/templates, and are
+			// ingested without going through the Pulumi Cloud.
+			continue
+		}
+
+		t := registryTemplate{template, r, s}
+		if templateName != "" && t.Name() != templateName {
+			continue
+		}
+
+		s.addTemplate(t)
+	}
+}
+
+type registryTemplate struct {
+	t        apitype.TemplateMetadata
+	registry registry.Registry
+	source   *Source
+}
+
+func (r registryTemplate) Name() string {
+	switch r.t.Source {
+	case "github", "gitlab":
+		parts := strings.SplitN(r.t.Name, "/", 3)
+		return parts[len(parts)-1]
+	default:
+		return r.t.Name
+	}
+}
+
+func (r registryTemplate) Description() string {
+	return ""
+}
+
+func (r registryTemplate) ProjectDescription() string {
+	if r.t.Description == nil {
+		return ""
+	}
+	return *r.t.Description
+}
+
+func (r registryTemplate) Error() error { return nil }
```

Comment:
Why is this needed?  Seems a bit strange to have an `Error()` method that always just returns `nil`?